tests/test_optimizer/test_dist_adafactor.py

```diff
@@ -0,0 +1,527 @@
+    weight_row_shard = shard_rowwise(weight.clone(), device_mesh.get_process_group(axis=1))
+    weight_row_shard_layout = get_layout(weight_row_shard) # Layout info weight_row_shard_layout.global_shape
+    weight_row_shard_shard_spec = get_sharding_spec(weight_row_shard) # Shard spec
+    weight_row_shard_flatten = nn.Parameter(weight_row_shard.clone().flatten().requires_grad_(True)) # flatten input(not dtensor) to optimizer
+    bias_row_flatten = nn.Parameter(bias.clone().flatten().requires_grad_(True))
+    row_params_shape = {id(weight_row_shard_flatten): weight_row_shard_layout.global_shape, id(bias_row_flatten): bias.shape} 
+    row_sharding_spec_dict = {id(weight_row_shard_flatten): weight_row_shard_shard_spec, id(bias_row_flatten): None}
+
+    # ==============================
+    # Init Optimizer 
+    # ==============================
+    
+    # base 
+    optimizer_base = Adafactor([weight, bias])
+    
+    # col parallel
+    optimizer_cp = DistributedAdaFactor([weight_col_shard_flatten, bias_col_flatten])
+    optimizer_cp.setup_distributed(tensor_parallel_group=tp_group, data_parallel_group=None, sharding_spec_dict=col_sharding_spec_dict, param_shape = col_params_shape)
+    # row parallel
+    optimizer_rp = DistributedAdaFactor([weight_row_shard_flatten, bias_row_flatten])
+    optimizer_rp.setup_distributed(tensor_parallel_group=tp_group, data_parallel_group =None, sharding_spec_dict=row_sharding_spec_dict, param_shape = row_params_shape)
+    
+    N_STEPS = 1
+    for _ in range(N_STEPS):
+        # base step
+        optimizer_base.zero_grad()
+        weight.grad = torch.rand_like(weight)
+        bias.grad = torch.rand_like(bias)
+        optimizer_base.step()
+        
+        # col parallel step
+        optimizer_cp.zero_grad()
+        weight_col_shard_flatten.grad = distribute_tensor(weight.grad, device_mesh, weight_col_shard_shard_spec).clone().flatten()
+        bias_col_flatten.grad = bias.grad.clone().flatten()
+        optimizer_cp.step()
+        
+        # row parallel step
+        optimizer_rp.zero_grad()
+        weight_row_shard_flatten.grad = distribute_tensor(weight.grad, device_mesh, weight_row_shard_shard_spec).clone().flatten()
+        bias_row_flatten.grad = bias.grad.clone().flatten()
+        optimizer_rp.step()
+        
+        # gather result
+        weight_col_gather = _gather(input_=weight_col_shard_flatten.data.view(-1, H // tensor_parallel_size),dim=-1, process_group=device_mesh.get_process_group(axis=1)) # gather
+        weight_row_gather = _gather(input_=weight_row_shard_flatten.data,dim=-1, process_group=device_mesh.get_process_group(axis=1)).view(-1, W) # gather
+        
+        
+        # verify
+        col_correct = correctness_verify(weight.data, weight_col_gather.data, dtype)
+        row_correct = correctness_verify(weight.data, weight_row_gather.data, dtype)
+        
+        print(f"col corrness {col_correct}  row correct {row_correct}")
```

Comment:
You may also check optimizer states as they may be sharded

Reply:
> You may also check optimizer states as they may be sharded


In fact, in the dist optim, some in-place operations are used on the state to save space, e.g. exp_avg_sq_col.div_(self.tensor_parallel_size); Therefore the optimizer state maybe different between the base optim and dist optim;